java.util.concurrent.TimeoutException: Connection lease request time out

Hello team,

I wanted to reach out reporting an issue 100% reproducible with reactive spring data elasticsearch please.

The setup is simple:

```
<parent>
        <groupId>org.springframework.boot</groupId>
        <artifactId>spring-boot-starter-parent</artifactId>
        <version>3.1.0-M1</version>
        <relativePath/>
    </parent>

       <dependency>
            <groupId>org.springframework.cloud</groupId>
            <artifactId>spring-cloud-stream-binder-kafka</artifactId>
            <version>4.0.2-SNAPSHOT</version>
        </dependency>
        <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-starter-actuator</artifactId>
        </dependency>
        <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-starter-data-elasticsearch</artifactId>
        </dependency>
```

```
@Configuration
@EnableReactiveElasticsearchRepositories(Constant.BASE_PACKAGE_ELASTIC)
public class ElasticConfiguration extends ReactiveElasticsearchConfiguration {

    @Override
    public final ClientConfiguration clientConfiguration() {
        return ClientConfiguration.builder().connectedTo("1xx.xx.xx.xx:9200").build();
    }

}

@Repository
public interface ElasticRepository extends ReactiveElasticsearchRepository<MyPojo, String> {
}

@org.springframework.data.elasticsearch.annotations.Document(indexName = Constant.ELASTIC_INDEX_NAME)
public record MyPojo(String name, String protocol, @Id String clientTs) {
}

```

The business logic is straightforward, using Reactor-Kafka; Spring Cloud Stream Kafka Reactive, or Spring Webflux, I will receive a flux of MyPojo. My goal is just to save them inside Elastic.

The rate is quite fast, therefore, I am using the reactive repository.

When using above setup, I will get 100%:

```
java.lang.RuntimeException: Connection lease request time out
	at org.springframework.data.elasticsearch.client.elc.ReactiveElasticsearchTemplate.translateException(ReactiveElasticsearchTemplate.java:670)
	at reactor.core.publisher.Flux.lambda$onErrorMap$27(Flux.java:7099)
	at reactor.core.publisher.FluxOnErrorResume$ResumeSubscriber.onError(FluxOnErrorResume.java:94)
	at reactor.core.publisher.MonoCompletionStage$MonoCompletionStageSubscription.apply(MonoCompletionStage.java:117)
	at reactor.core.publisher.MonoCompletionStage$MonoCompletionStageSubscription.apply(MonoCompletionStage.java:64)
	at java.base/java.util.concurrent.CompletableFuture.uniHandle(CompletableFuture.java:934)
	at java.base/java.util.concurrent.CompletableFuture$UniHandle.tryFire(CompletableFuture.java:911)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2162)
	at co.elastic.clients.transport.rest_client.RestClientTransport$1.onFailure(RestClientTransport.java:179)
	at org.elasticsearch.client.RestClient$FailureTrackingResponseListener.onDefinitiveFailure(RestClient.java:684)
	at org.elasticsearch.client.RestClient$1.failed(RestClient.java:422)
	at org.apache.http.concurrent.BasicFuture.failed(BasicFuture.java:137)
	at org.apache.http.impl.nio.client.DefaultClientExchangeHandlerImpl.executionFailed(DefaultClientExchangeHandlerImpl.java:101)
	at org.apache.http.impl.nio.client.AbstractClientExchangeHandler.failed(AbstractClientExchangeHandler.java:432)
	at org.apache.http.impl.nio.client.AbstractClientExchangeHandler.connectionRequestFailed(AbstractClientExchangeHandler.java:352)
	at org.apache.http.impl.nio.client.AbstractClientExchangeHandler.access$100(AbstractClientExchangeHandler.java:64)
	at org.apache.http.impl.nio.client.AbstractClientExchangeHandler$1.failed(AbstractClientExchangeHandler.java:396)
	at org.apache.http.concurrent.BasicFuture.failed(BasicFuture.java:137)
	at org.apache.http.impl.nio.conn.PoolingNHttpClientConnectionManager$1.failed(PoolingNHttpClientConnectionManager.java:318)
	at org.apache.http.concurrent.BasicFuture.failed(BasicFuture.java:137)
	at org.apache.http.nio.pool.AbstractNIOConnPool.fireCallbacks(AbstractNIOConnPool.java:501)
	at org.apache.http.nio.pool.AbstractNIOConnPool.release(AbstractNIOConnPool.java:360)
	at org.apache.http.impl.nio.conn.PoolingNHttpClientConnectionManager.releaseConnection(PoolingNHttpClientConnectionManager.java:393)
	at org.apache.http.impl.nio.client.AbstractClientExchangeHandler.releaseConnection(AbstractClientExchangeHandler.java:247)
	at org.apache.http.impl.nio.client.MainClientExec.responseCompleted(MainClientExec.java:387)
	at org.apache.http.impl.nio.client.DefaultClientExchangeHandlerImpl.responseCompleted(DefaultClientExchangeHandlerImpl.java:173)
	at org.apache.http.nio.protocol.HttpAsyncRequestExecutor.processResponse(HttpAsyncRequestExecutor.java:448)
	at org.apache.http.nio.protocol.HttpAsyncRequestExecutor.inputReady(HttpAsyncRequestExecutor.java:338)
	at org.apache.http.impl.nio.DefaultNHttpClientConnection.consumeInput(DefaultNHttpClientConnection.java:265)
	at org.apache.http.impl.nio.client.InternalIODispatch.onInputReady(InternalIODispatch.java:87)
	at org.apache.http.impl.nio.client.InternalIODispatch.onInputReady(InternalIODispatch.java:40)
	at org.apache.http.impl.nio.reactor.AbstractIODispatch.inputReady(AbstractIODispatch.java:121)
	at org.apache.http.impl.nio.reactor.BaseIOReactor.readable(BaseIOReactor.java:162)
	at org.apache.http.impl.nio.reactor.AbstractIOReactor.processEvent(AbstractIOReactor.java:337)
	at org.apache.http.impl.nio.reactor.AbstractIOReactor.processEvents(AbstractIOReactor.java:315)
	at org.apache.http.impl.nio.reactor.AbstractIOReactor.execute(AbstractIOReactor.java:276)
	at org.apache.http.impl.nio.reactor.BaseIOReactor.execute(BaseIOReactor.java:104)
	at org.apache.http.impl.nio.reactor.AbstractMultiworkerIOReactor$Worker.run(AbstractMultiworkerIOReactor.java:591)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.util.concurrent.TimeoutException: Connection lease request time out
	at org.apache.http.nio.pool.AbstractNIOConnPool.processPendingRequest(AbstractNIOConnPool.java:411)
	at org.apache.http.nio.pool.AbstractNIOConnPool.processNextPendingRequest(AbstractNIOConnPool.java:391)
	at org.apache.http.nio.pool.AbstractNIOConnPool.release(AbstractNIOConnPool.java:355)
	... 17 common frames omitted
```

The thing is:
I tried ``` .withConnectTimeout() .withSocketTimeout()`` set to some crazy value, still getting this error.

I tried switching to non reactive, and it works.

Finally, I will see this stack trace after processing many messages. Yet, I will see only 1 Hit in Elasticsearch.

May I ask what is the root cause of this issue please?

Thank you

--- a/src/main/java/org/springframework/data/elasticsearch/core/AbstractReactiveElasticsearchTemplate.java
+++ b/src/main/java/org/springframework/data/elasticsearch/core/AbstractReactiveElasticsearchTemplate.java
@@ -17,13 +17,17 @@ package org.springframework.data.elasticsearch.core;
 
 import reactor.core.publisher.Flux;
 import reactor.core.publisher.Mono;
+import reactor.core.publisher.Sinks;
 import reactor.util.function.Tuple2;
 
 import java.time.Duration;
 import java.util.Collection;
 import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.stream.Collectors;
 
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import org.springframework.beans.BeansException;
@@ -205,6 +209,61 @@ abstract public class AbstractReactiveElasticsearchTemplate
 	@Override
 	public <T> Mono<T> save(T entity) {
 		return save(entity, getIndexCoordinatesFor(entity.getClass()));
+	}
+
+	@Override
+	public <T> Flux<T> save(Flux<T> entities, Class<?> clazz, int bulkSize) {
+		return save(entities, getIndexCoordinatesFor(clazz), bulkSize);
+	}
+
+	@Override
+	public <T> Flux<T> save(Flux<T> entities, IndexCoordinates index, int bulkSize) {
+
+		Assert.notNull(entities, "entities must not be null");
+		Assert.notNull(index, "index must not be null");
+		Assert.isTrue(bulkSize > 0, "bulkSize must be greater than 0");
+
+		return Flux.defer(() -> {
+			Sinks.Many<T> sink = Sinks.many().unicast().onBackpressureBuffer();
+			entities //
+					.bufferTimeout(bulkSize, Duration.ofMillis(200)) //
+					.subscribe(new Subscriber<List<T>>() {
+						private Subscription subscription;
+						private AtomicBoolean upstreamComplete = new AtomicBoolean(false);
+
+						@Override
+						public void onSubscribe(Subscription subscription) {
+							this.subscription = subscription;
+							subscription.request(1);
+						}
+
+						@Override
+						public void onNext(List<T> entityList) {
+							saveAll(entityList, index) //
+									.map(sink::tryEmitNext) //
+									.doOnComplete(() -> {
+										if (!upstreamComplete.get()) {
+											subscription.request(1);
+										} else {
+											sink.tryEmitComplete();
+										}
+									}).subscribe();
+						}
+
+						@Override
+						public void onError(Throwable throwable) {
+							subscription.cancel();
+							sink.tryEmitError(throwable);
+						}
+
+						@Override
+						public void onComplete() {
+							upstreamComplete.set(true);
+						}
+					});
+			return sink.asFlux();
+		});
+
 	}
 
 	@Override

--- a/src/main/java/org/springframework/data/elasticsearch/core/ReactiveDocumentOperations.java
+++ b/src/main/java/org/springframework/data/elasticsearch/core/ReactiveDocumentOperations.java
@@ -44,6 +44,9 @@ import org.springframework.util.Assert;
  * @since 4.0
  */
 public interface ReactiveDocumentOperations {
+
+	int FLUX_SAVE_BULK_SIZE = 500;
+
 	/**
 	 * Index the given entity, once available, extracting index from entity metadata.
 	 *
@@ -92,6 +95,61 @@ public interface ReactiveDocumentOperations {
 	 * @return a {@link Mono} emitting the saved entity.
 	 */
 	<T> Mono<T> save(T entity, IndexCoordinates index);
+
+	/**
+	 * Indexes the entities into the index extracted from entity metadata.
+	 *
+	 * @param entities
+	 * @param clazz the class to get the index name from
+	 * @param <T> entity type
+	 * @return a Flux emitting the saved entities
+	 * @since 5.1
+	 */
+	default <T> Flux<T> save(Flux<T> entities, Class<?> clazz) {
+		return save(entities, clazz, FLUX_SAVE_BULK_SIZE);
+	}
+
+	/**
+	 * Indexes the entities into the index extracted from entity metadata. The entities are collected into batches of
+	 * {bulkSize} with a maximal timeout of 200 ms, see
+	 * {@link reactor.core.publisher.Flux#bufferTimeout(int, java.time .Duration)} and then sent in a bulk operation to
+	 * Elasticsearch.
+	 *
+	 * @param entities
+	 * @param clazz the class to get the index name from
+	 * @param bulkSize number of entities to put in a bulk request
+	 * @param <T> entity type
+	 * @return a Flux emitting the saved entities
+	 * @since 5.1
+	 */
+	<T> Flux<T> save(Flux<T> entities, Class<?> clazz, int bulkSize);
+
+	/**
+	 * Indexes the entities into the given index.
+	 *
+	 * @param entities the entities to save
+	 * @param index the index to save to
+	 * @param <T> entity type
+	 * @return a Flux emitting the saved entities
+	 * @since 5.1
+	 */
+	default <T> Flux<T> save(Flux<T> entities, IndexCoordinates index) {
+		return save(entities, index, FLUX_SAVE_BULK_SIZE);
+	}
+
+	/**
+	 * Indexes the entities into the given index. The entities are collected into batches of {bulkSize} with a maximal
+	 * timeout of 200 ms, see {@link reactor.core.publisher.Flux#bufferTimeout(int, java.time * .Duration)} and then sent
+	 * in a bulk operation to Elasticsearch.
+	 *
+	 * @param entities the entities to save
+	 * @param index the index to save to
+	 * @param bulkSize number of entities to put in a bulk request
+	 * @param <T> entity type
+	 * @return a Flux emitting the saved entities
+	 * @since 5.1
+	 */
+	<T> Flux<T> save(Flux<T> entities, IndexCoordinates index, int bulkSize);
 
 	/**
 	 * Index entities the index extracted from entity metadata.

--- a/src/main/java/org/springframework/data/elasticsearch/repository/support/SimpleReactiveElasticsearchRepository.java
+++ b/src/main/java/org/springframework/data/elasticsearch/repository/support/SimpleReactiveElasticsearchRepository.java
@@ -15,7 +15,6 @@
  */
 package org.springframework.data.elasticsearch.repository.support;
 
-import org.springframework.data.elasticsearch.core.query.BaseQuery;
 import reactor.core.publisher.Flux;
 import reactor.core.publisher.Mono;
 
@@ -30,6 +29,7 @@ import org.springframework.data.elasticsearch.core.RefreshPolicy;
 import org.springframework.data.elasticsearch.core.SearchHit;
 import org.springframework.data.elasticsearch.core.mapping.ElasticsearchPersistentEntity;
 import org.springframework.data.elasticsearch.core.mapping.IndexCoordinates;
+import org.springframework.data.elasticsearch.core.query.BaseQuery;
 import org.springframework.data.elasticsearch.core.query.Query;
 import org.springframework.data.elasticsearch.repository.ReactiveElasticsearchRepository;
 import org.springframework.util.Assert;
@@ -97,7 +97,7 @@ public class SimpleReactiveElasticsearchRepository<T, ID> implements ReactiveEla
 
 		Assert.notNull(entityStream, "EntityStream must not be null!");
 
-		return operations.saveAll(Flux.from(entityStream).collectList(), entityInformation.getIndexCoordinates())
+		return operations.save(Flux.from(entityStream), entityInformation.getIndexCoordinates())
 				.concatWith(doRefresh().then(Mono.empty()));
 	}
 

--- a/src/test/java/org/springframework/data/elasticsearch/core/ReactiveElasticsearchIntegrationTests.java
+++ b/src/test/java/org/springframework/data/elasticsearch/core/ReactiveElasticsearchIntegrationTests.java
@@ -20,7 +20,7 @@ import static org.assertj.core.api.Assertions.*;
 import static org.elasticsearch.index.query.QueryBuilders.*;
 import static org.springframework.data.elasticsearch.annotations.FieldType.*;
 
-import org.springframework.data.elasticsearch.annotations.IndexedIndexName;
+import reactor.core.publisher.Flux;
 import reactor.core.publisher.Mono;
 import reactor.test.StepVerifier;
 
@@ -61,6 +61,7 @@ import org.springframework.data.elasticsearch.RestStatusException;
 import org.springframework.data.elasticsearch.annotations.Document;
 import org.springframework.data.elasticsearch.annotations.Field;
 import org.springframework.data.elasticsearch.annotations.FieldType;
+import org.springframework.data.elasticsearch.annotations.IndexedIndexName;
 import org.springframework.data.elasticsearch.annotations.Mapping;
 import org.springframework.data.elasticsearch.annotations.Setting;
 import org.springframework.data.elasticsearch.annotations.WriteOnlyProperty;
@@ -169,7 +170,6 @@ public abstract class ReactiveElasticsearchIntegrationTests {
 
 		assertThat(saved.getIndexedIndexName()).isEqualTo(indexNameProvider.indexName() + "-indexedindexname");
 	}
-
 
 	private Mono<Boolean> documentWithIdExistsInIndex(String id, String index) {
 		return operations.exists(id, IndexCoordinates.of(index));
@@ -1180,6 +1180,25 @@ public abstract class ReactiveElasticsearchIntegrationTests {
 					assertThat(readEntity.getPart2()).isEqualTo(entity.getPart2()); //
 				}).verifyComplete();
 	}
+
+	@Test // #2496
+	@DisplayName("should save data from Flux and return saved data in a flux")
+	void shouldSaveDataFromFluxAndReturnSavedDataInAFlux() {
+
+		var count = 12_345;
+		var entityList = IntStream.rangeClosed(1, count)//
+				.mapToObj(SampleEntity::of) //
+				.collect(Collectors.toList());
+
+		var entityFlux = Flux.fromIterable(entityList);
+
+		operations.save(entityFlux, SampleEntity.class).collectList() //
+				.as(StepVerifier::create) //
+				.consumeNextWith(savedEntities -> {
+					assertThat(savedEntities).isEqualTo(entityList);
+				}) //
+				.verifyComplete();
+	}
 	// endregion
 
 	// region Helper functions
@@ -1261,6 +1280,13 @@ public abstract class ReactiveElasticsearchIntegrationTests {
 		@Nullable private int rate;
 		@Nullable
 		@Version private Long version;
+
+		static SampleEntity of(int id) {
+			var entity = new SampleEntity();
+			entity.setId("" + id);
+			entity.setMessage(" message " + id);
+			return entity;
+		}
 
 		@Nullable
 		public String getId() {
@@ -1543,6 +1569,7 @@ public abstract class ReactiveElasticsearchIntegrationTests {
 			this.part2 = part2;
 		}
 	}
+
 	@Document(indexName = "#{@indexNameProvider.indexName()}-indexedindexname")
 	private static class IndexedIndexNameEntity {
 		@Nullable
@@ -1550,8 +1577,7 @@ public abstract class ReactiveElasticsearchIntegrationTests {
 		@Nullable
 		@Field(type = Text) private String someText;
 		@Nullable
-		@IndexedIndexName
-		private String indexedIndexName;
+		@IndexedIndexName private String indexedIndexName;
 
 		@Nullable
 		public String getId() {
@@ -1579,5 +1605,6 @@ public abstract class ReactiveElasticsearchIntegrationTests {
 		public void setIndexedIndexName(@Nullable String indexedIndexName) {
 			this.indexedIndexName = indexedIndexName;
 		}
-	}	// endregion
+	}
+	// endregion
 }

--- a/src/test/java/org/springframework/data/elasticsearch/core/suggest/CompletionIntegrationTests.java
+++ b/src/test/java/org/springframework/data/elasticsearch/core/suggest/CompletionIntegrationTests.java
@@ -112,7 +112,6 @@ public abstract class CompletionIntegrationTests implements NewElasticsearchClie
 		operations.bulkIndex(indexQueries, AnnotatedCompletionEntity.class);
 	}
 
-	// @DisabledIf(value = "newElasticsearchClient", disabledReason = "todo #2139, ES issue 150")
 	@Test
 	public void shouldFindSuggestionsForGivenCriteriaQueryUsingCompletionEntity() {
 
@@ -144,7 +143,6 @@ public abstract class CompletionIntegrationTests implements NewElasticsearchClie
 		operations.get("1", CompletionEntity.class);
 	}
 
-	// @DisabledIf(value = "newElasticsearchClient", disabledReason = "todo #2139, ES issue 150")
 	@Test
 	public void shouldFindSuggestionsForGivenCriteriaQueryUsingAnnotatedCompletionEntity() {
 
@@ -168,7 +166,6 @@ public abstract class CompletionIntegrationTests implements NewElasticsearchClie
 		assertThat(options.get(1).getText()).isIn("Marchand", "Mohsin");
 	}
 
-	// @DisabledIf(value = "newElasticsearchClient", disabledReason = "todo #2139, ES 1issue 150")
 	@Test
 	public void shouldFindSuggestionsWithWeightsForGivenCriteriaQueryUsingAnnotatedCompletionEntity() {
 

--- a/src/test/java/org/springframework/data/elasticsearch/core/suggest/ReactiveSuggestIntegrationTests.java
+++ b/src/test/java/org/springframework/data/elasticsearch/core/suggest/ReactiveSuggestIntegrationTests.java
@@ -66,7 +66,6 @@ public abstract class ReactiveSuggestIntegrationTests implements NewElasticsearc
 		operations.indexOps(IndexCoordinates.of(indexNameProvider.getPrefix() + "*")).delete().block();
 	}
 
-	// @DisabledIf(value = "newElasticsearchClient", disabledReason = "todo #2139, ES issue 150")
 	@Test // #1302
 	@DisplayName("should find suggestions for given prefix completion")
 	void shouldFindSuggestionsForGivenPrefixCompletion() {

--- a/src/test/java/org/springframework/data/elasticsearch/repository/support/SimpleReactiveElasticsearchRepositoryIntegrationTests.java
+++ b/src/test/java/org/springframework/data/elasticsearch/repository/support/SimpleReactiveElasticsearchRepositoryIntegrationTests.java
@@ -739,6 +739,24 @@ abstract class SimpleReactiveElasticsearchRepositoryIntegrationTests {
 				.verifyComplete();
 	}
 
+	@Test // #2496
+	@DisplayName("should save data from Flux and return saved data in a flux")
+	void shouldSaveDataFromFluxAndReturnSavedDataInAFlux() {
+		var count = 12_345;
+		var entityList = IntStream.rangeClosed(1, count)//
+				.mapToObj(SampleEntity::of) //
+				.collect(Collectors.toList());
+
+		var entityFlux = Flux.fromIterable(entityList);
+
+		repository.saveAll(entityFlux).collectList() //
+				.as(StepVerifier::create) //
+				.consumeNextWith(savedEntities -> {
+					assertThat(savedEntities).isEqualTo(entityList);
+				}) //
+				.verifyComplete();
+	}
+
 	Mono<Void> bulkIndex(SampleEntity... entities) {
 		return operations.saveAll(Arrays.asList(entities), IndexCoordinates.of(indexNameProvider.indexName())).then();
 	}
@@ -828,6 +846,13 @@ abstract class SimpleReactiveElasticsearchRepositoryIntegrationTests {
 		@Version private Long version;
 		@Field(name = "custom_field_name", type = FieldType.Text)
 		@Nullable private String customFieldNameMessage;
+
+		static SampleEntity of(int id) {
+			var entity = new SampleEntity();
+			entity.setId("" + id);
+			entity.setMessage(" message " + id);
+			return entity;
+		}
 
 		public SampleEntity() {}
 

